source/common/filesystem/win32/filesystem_impl.cc

```diff
@@ -102,8 +104,120 @@ std::string InstanceImplWin32::fileReadToEnd(const std::string& path) {
+    1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0,
+ // @  A  B  C  D  E  F  G  H  I  J  K  L  M  N  O  P  Q  R  S  T  U  V  W  X  Y  Z  [  \  ]  ^  _
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
+ // `  a  b  c  d  e  f  g  h  i  j  k  l  m  n  o  p  q  r  s  t  u  v  w  x  y  z  {  |  }  ~
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,
+ // High bit codes are accepted (subject to code page translation rules)
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
+
+// The "COM#" and "LPT#" names below have boolean flag requiring a [1-9] suffix.
+// This list can be avoided by observing dwFileAttributes & FILE_ATTRIBUTE_DEVICE
+// within WIN32_FILE_ATTRIBUTE_DATA or WIN32_FIND_DATA results.
+std::unordered_map<std::string, bool> pathelt_table = {
+    {"CON", false}, {"NUL", false}, {"AUX", false}, {"PRN", false}, {"COM", true}, {"LPT", true}};
+
+// clang-format on
+
 bool InstanceImplWin32::illegalPath(const std::string& path) {
-  // Currently, we don't know of any obviously illegal paths on Windows
+  std::string pathbuffer = path;
+  absl::string_view pathname = pathbuffer;
+
+  // Examine and skip common leading path patterns of \\?\ and
+  // reject paths with any other leading \\.\ device or an
+  // unrecognized \\*\ prefix
+  if ((pathname.size() >= 4) && (pathname[0] == '/' || pathname[0] == '\\') &&
+      (pathname[1] == '/' || pathname[1] == '\\') && (pathname[3] == '/' || pathname[3] == '\\')) {
+    if (pathname[2] == '?')
+      pathname = pathname.substr(4);
+    else
+      return true;
+  }
+  // Examine and accept D: drive prefix (last opportunity to
+  // accept a colon in the file path) and skip the D: component
+  // This may result in a relative-to working directory or absolute path on D:
+  if (pathname.size() >= 2 && std::isalpha(pathname[0]) && pathname[1] == ':') {
+    pathname = pathname.substr(2);
+  }
+  std::string ucase_prefix("   ");
+  std::vector<std::string> pathelts = absl::StrSplit(pathname, absl::ByAnyChar("/\\"));
+  for (const std::string& elt : pathelts) {
+    // Accept element of empty, ".", ".." as special cases,
+    if (elt.size() == 0 ||
+        (elt[0] == '.' && (elt.size() == 1 || (elt[1] == '.' && (elt.size() == 2)))))
+      continue;
+    // Upper-case path segment prefix to compare to character device names
+    if (elt.size() >= 3) {
+      int i;
+      for (i = 0; i < 3; ++i)
+        ucase_prefix[i] = ::toupper(elt[i]);
```

Comment:
check everything for braces

Reply:
Happy to re-review and submit further fixes once the automation is in place, thanks for clarifying
the style convention.